utils/key-gen/key-gen.cpp

```diff
@@ -0,0 +1,330 @@
+    if (paramsOpts.mvec.length() == 0) {
+      std::cerr << "Missing mvec parameter for bootstrapping in "
+                << cmdLineOpts.paramFileName << "." << std::endl;
+      return EXIT_FAILURE;
+    }
+    if (paramsOpts.gens.length() == 0) {
+      std::cerr << "Missing gens parameter for bootstrapping in "
+                << cmdLineOpts.paramFileName << "." << std::endl;
+      return EXIT_FAILURE;
+    }
+    if (paramsOpts.ords.length() == 0) {
+      std::cerr << "Missing ords parameter for bootstrapping in "
+                << cmdLineOpts.paramFileName << "." << std::endl;
+      return EXIT_FAILURE;
+    }
+  }
+
+  try {
+    helib::Context* contextp;
+
+    if (cmdLineOpts.scheme == "BGV") {
+      helib::ContextBuilder<helib::BGV> cb;
+      cb.m(paramsOpts.m)
+          .p(p)
+          .r(paramsOpts.r)
+          .gens(helib::convert<std::vector<long>>(paramsOpts.gens))
+          .ords(helib::convert<std::vector<long>>(paramsOpts.ords))
+          .bits(paramsOpts.Qbits)
+          .c(paramsOpts.c);
+
+      if (cmdLineOpts.bootstrappable != "NONE") {
+        cb.bootstrappable(true).mvec(paramsOpts.mvec);
+        if (cmdLineOpts.bootstrappable == "THICK")
+          cb.thickboot();
+        else if (cmdLineOpts.bootstrappable == "THIN")
+          cb.thinboot();
+      }
+      contextp = cb.buildPtr();
+    } else if (cmdLineOpts.scheme == "CKKS") {
+      contextp = helib::ContextBuilder<helib::CKKS>()
+                     .m(paramsOpts.m)
+                     .precision(paramsOpts.r)
+                     .bits(paramsOpts.Qbits)
+                     .scale(paramsOpts.scale)
+                     .buildPtr();
+    }
+
+    // and a new secret/public key
+    helib::SecKey secretKey(*contextp);
+    secretKey.GenSecKey(); // A +-1/0 secret key
+
+    // If not set by user, returns params file name with truncated UTC
+    if (cmdLineOpts.outputPrefixPath.empty()) {
+      cmdLineOpts.outputPrefixPath =
+          stripExtension(cmdLineOpts.paramFileName) +
+          std::to_string(std::time(nullptr) % 100000);
+      std::cout << "File prefix: " << cmdLineOpts.outputPrefixPath << std::endl;
+    }
+
+    // Printout important info
+    if (cmdLineOpts.infoFile) {
+      // outputPrefixPath should be set further up main.
+      std::string path = cmdLineOpts.outputPrefixPath + ".info";
+      std::ofstream out(path);
+      if (!out.is_open()) {
+        throw std::runtime_error("Cannot write keys to file at '" + path +
+                                 "'.");
+      }
+      printoutToStream(*contextp,
+                       out,
+                       cmdLineOpts.noSKM,
+                       cmdLineOpts.frobSKM,
+                       cmdLineOpts.bootstrappable != "NONE");
+    } else {
+      printoutToStream(*contextp,
+                       std::cout,
+                       cmdLineOpts.noSKM,
+                       cmdLineOpts.frobSKM,
+                       cmdLineOpts.bootstrappable != "NONE");
+    }
```

Comment:
The only difference to the `printoutToStream` call is the stream that is being passed. Declare pointer to `std::ostream` then use the if to set the `std::ostream` to either `std::cout` or create a `std::ofstream`. Then you only have to call the function in one way.